plum/parametric.py

```diff
@@ -293,3 +325,40 @@ def type_of(obj: tuple):
 # Deliver `type_of`.
 promised_type_of1.deliver(type_of)
 promised_type_of2.deliver(type_of)
+
+
+@parametric
+class Val:
+    """
+    A parametric type used to move information from the value-domain to the type-domain.
+    """
+
+    @classmethod
+    def __infer_type_parameter__(cls, *arg, **kwargs):
+        """Function called when the constructor of `Val` is called to determine the type
+        parameters.
+        """
+        if len(arg) == 0:
+            raise ValueError("The value must be specified.")
+        elif len(arg) > 1:
+            raise ValueError("Too many values. Val accepts only one argument.")
+        return arg[0]
+
+    def __init__(self, arg=None):
+        """
+        Construct a Value object with type `Val(arg)` that cna be used to dispatch
+        based on values.
+
+        Args:
+            arg: the value to be moved to the type domain
+        """
+        if type(self).is_concrete:
+            return
+        else:
+            raise ValueError("The value must be specified.")  # pragma: no cover
+
+    def __repr__(self):
+        return "{}()".format(repr(type(self)))
+
+    def __eq__(self, other):
```

Comment:
```suggestion
    @_dispatch
    def __eq__(self, other: "Val"):
```

Maybe add this for safety? Perhaps also add a quick test to test this?

Reply:
Why? 
I think this is already pretty safe, and avoids an extra dispatch...
Unless we want to make this extensible (but do we?)

Reply:
Hmm, yeah, I was worried about the case where `other` isn't a `Val`, but checking the types should catch that.

Reply:
What do you mean by checking the types?

Reply:
The implementation  of `__eq__` is `type(self) == type(other)`, so that can never be true if `other` isn't a `Val`, because then the type of `other` will be different. Think this looks good. :)